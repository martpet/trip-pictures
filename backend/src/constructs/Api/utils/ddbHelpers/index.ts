export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
