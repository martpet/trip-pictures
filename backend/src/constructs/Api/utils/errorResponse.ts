import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { ApiErrorResponseBody } from '~/types';

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

export const errorResponse = (
  traceId: string,
  options: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  } = options;

  const { envName } = globalLambdaProps;

  const errObj: ApiErrorResponseBody['error'] = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errObj.description = description;
  }

  if (error && (exposeError ?? envName !== 'production')) {
    errObj.error = String(error);
  }

  const body: ApiErrorResponseBody = {
    error: errObj,
  };

  console.error({ traceId, error });

  return {
    statusCode,
    body: JSON.stringify(body),
  };
};
