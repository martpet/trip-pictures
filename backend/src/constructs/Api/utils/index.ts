export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
