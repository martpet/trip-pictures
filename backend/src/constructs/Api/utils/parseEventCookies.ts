import { APIGatewayProxyEventV2 } from 'aws-lambda';
import cookie from 'cookie';

import { Cookies, PathWithCookies } from '../types';

export const parseEventCookies = <T extends PathWithCookies>(
  event: APIGatewayProxyEventV2
): Cookies<T> => {
  try {
    const cookieHeaderString = event.cookies?.join(';');
    if (!cookieHeaderString) throw new Error();
    return cookie.parse(cookieHeaderString);
  } catch (e) {
    return {};
  }
};
