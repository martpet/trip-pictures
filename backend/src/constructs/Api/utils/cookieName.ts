import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
