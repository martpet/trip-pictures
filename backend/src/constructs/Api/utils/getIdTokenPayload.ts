import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = async (idToken: string): Promise<IdTokenPayload> => {
  const camelcaseKeys = (await import('camelcase-keys')).default;
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
