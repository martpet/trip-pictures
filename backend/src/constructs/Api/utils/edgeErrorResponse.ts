import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
