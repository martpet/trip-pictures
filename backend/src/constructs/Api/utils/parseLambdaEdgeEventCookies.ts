import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { LambdaEdgeViewerEvent } from '~/constructs/Api/types';
import { parseEventCookies } from '~/constructs/Api/utils/parseEventCookies';

export const parseLambdaEdgeEventCookies = (edgeEvent: LambdaEdgeViewerEvent) => {
  const { request } = edgeEvent.Records[0].cf;
  const cookies = request.headers.cookie?.map(({ value }) => value);
  const apiGatewayEvent = { cookies } as unknown as APIGatewayProxyEventV2;

  return parseEventCookies(apiGatewayEvent);
};
