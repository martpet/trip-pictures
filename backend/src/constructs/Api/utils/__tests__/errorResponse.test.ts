import { StatusCodes } from 'http-status-codes';

import { errorResponse, itCalls, itReturns } from '~/constructs/Api/utils';
import { EnvName } from '~/types';

const args: Parameters<typeof errorResponse> = ['dummyTraceId'];

describe('errorResponse', () => {
  itCalls(console.error, errorResponse, args);
  itReturns(errorResponse, args);

  describe('when "statusCode" is provided', () => {
    const argsClone = structuredClone(args);
    argsClone.push({ statusCode: StatusCodes.IM_A_TEAPOT });
    itReturns(errorResponse, argsClone);
  });

  describe('when "description" is provided', () => {
    const argsClone = structuredClone(args);
    argsClone.push({ description: 'dummDescription' });
    itReturns(errorResponse, argsClone);
  });

  describe('when "error" is provided', () => {
    const argsWithError = structuredClone(args) as Required<typeof args>;
    argsWithError.push({ error: new Error('dummyErrorMessage') });

    itCalls(console.error, errorResponse, argsWithError);

    describe('when "exposeError" is true', () => {
      const argsWithExposedError = structuredClone(argsWithError);
      argsWithExposedError[1].exposeError = true;
      itReturns(errorResponse, argsWithExposedError);
    });

    describe.each(['personal', 'staging'])(
      'when "globalLambdaProps.envName" is "%s"',
      (key) => {
        beforeEach(() => {
          globalLambdaProps.envName = key as EnvName;
        });
        afterEach(() => {
          globalLambdaProps.envName = 'production';
        });
        itReturns(errorResponse, argsWithError);
      }
    );
  });
});
