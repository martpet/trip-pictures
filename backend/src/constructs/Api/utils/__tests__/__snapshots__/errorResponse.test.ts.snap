// Vitest Snapshot v1

exports[`errorResponse > calls "console.error" with correct args 1`] = `
[
  [
    {
      "error": undefined,
      "traceId": "dummyTraceId",
    },
  ],
]
`;

exports[`errorResponse > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"traceId\\":\\"dummyTraceId\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "description" is provided > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"traceId\\":\\"dummyTraceId\\",\\"description\\":\\"dummDescription\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "error" is provided > calls "console.error" with correct args 1`] = `
[
  [
    {
      "error": [Error: dummyErrorMessage],
      "traceId": "dummyTraceId",
    },
  ],
]
`;

exports[`errorResponse > when "error" is provided > when "exposeError" is true > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"traceId\\":\\"dummyTraceId\\",\\"error\\":\\"Error: dummyErrorMessage\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "error" is provided > when "globalLambdaProps.envName" is "personal" > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"traceId\\":\\"dummyTraceId\\",\\"error\\":\\"Error: dummyErrorMessage\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "error" is provided > when "globalLambdaProps.envName" is "staging" > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"traceId\\":\\"dummyTraceId\\",\\"error\\":\\"Error: dummyErrorMessage\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "statusCode" is provided > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":418,\\"message\\":\\"I'm a teapot\\",\\"traceId\\":\\"dummyTraceId\\"}}",
  "statusCode": 418,
}
`;
